model/src/payouts/quanto.rs

```diff
@@ -0,0 +1,609 @@
+    let margin = Amount::from_btc(margin).context("Could not convert margin to bitcoin::Amount")?;
+
+    Ok(margin)
+}
+
+/// The profit and loss (PNL).
+///
+/// It is convenient to model the calculations for long and short together, because one party's gain
+/// is the other one's loss and vice-versa. That is, the absolute value of PNL will be the same,
+/// with only the sign changing between the two parties.
+struct Pnl(SignedAmount);
+
+impl Pnl {
+    /// Compute the PNL of the contract.
+    ///
+    /// Call `Self::new().long()` and `Self::new().short()` to access the PNL values for long and
+    /// short respectively.
+    fn new(
+        initial_price: u64,
+        closing_price: u64,
+        multiplier: Decimal,
+        n_contracts: u64,
+    ) -> Result<Self> {
+        let initial_price = Decimal::from(initial_price);
+        let closing_price = Decimal::from(closing_price);
+        let n_contracts = Decimal::from(n_contracts);
+
+        let pnl = (closing_price - initial_price) * multiplier * n_contracts;
+        let pnl = pnl.to_f64().context("Could not convert PNL to f64")?;
+        let pnl = SignedAmount::from_btc(pnl)
+            .context("Could not convert PNL to bitcoin::SignedAmount")?;
+
+        Ok(Self(pnl))
+    }
+
+    /// The profit and loss (PNL) from the perspective of the party going long.
+    fn long(&self) -> SignedAmount {
+        self.0
+    }
+
+    /// The profit and loss (PNL) from the perspective of the party going short.
+    fn short(&self) -> SignedAmount {
+        self.0 * -1
+    }
+}
+
+/// Closing prices at which parties get liquidated i.e. when they get a 0 payout.
+struct LiquidationThresholds {
+    inner: u64, // TODO: Find a better name!
+    initial_price: u64,
+}
+
+impl LiquidationThresholds {
+    /// Compute the liquidation thresholds for both parties.
+    ///
+    /// Call `Self::new().long()` and `Self::new().short()` to access the liquidation threshold
+    /// values for long and short respectively.
+    fn new(
+        initial_margin: Amount,
+        n_contracts: u64,
+        initial_price: u64,
+        multiplier: Decimal,
+    ) -> Result<Self> {
+        let initial_margin = Decimal::from_f64(initial_margin.as_btc())
+            .context("Could not create Decimal from initial margin")?;
+
+        let n_contracts = Decimal::from(n_contracts);
+
+        let inner = initial_margin / (multiplier * n_contracts);
+        let inner = inner
+            .to_u64()
+            .context("Could not convert inner threshold calculation to u64")?;
+
+        Ok(Self {
+            inner,
+            initial_price,
+        })
+    }
+
+    /// Compute the closing price under which the party going long gets liquidated.
+    fn long(&self) -> u64 {
+        self.initial_price.saturating_sub(self.inner)
+    }
+
+    /// Compute the closing price over which the party going short gets liquidated.
+    fn short(&self) -> u64 {
+        self.initial_price + self.inner
+    }
```

Comment:
Given `short` and `long` are based on constants within `LiquidationThresholds` , why don't you precompute them?

Reply:
I see your point. But each of them is only called once, separately. So it's a minuscule bit more efficient to compute them on demand.

The only reason why I introduced the type was so that they could share the code between them. Free functions would be valid but then I would have to return the intermediate values in a tuple, which is kind of ugly.